src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
                 {
-                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
                 }
 
-                // TODO: Handle requests with body
+                if (state.RequestMessage.Content != null)
+                {
+#if DEBUG
+                    Logging.PrintInfo(Logging.Http, this, "Request content for " + Logging.GetObjectLogHash(responseMessage.ContentStream) + " : " + state.RequestMessage.Content.ReadAsStringAsync().Result);
+#endif
+
+                    Stream requestContentStream =
+                        await state.RequestMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                    if (state.CancellationToken.IsCancellationRequested)
+                    {
+                        state.TrySetCanceled(state.CancellationToken);
+                        return;
+                    }
+                    state.RequestContentStream = requestContentStream;
+                }
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                requestHandle.DangerousAddRef(ref referenceTaken);
+                var performTask = Task.Run(() =>
+                    {
+                        if (state.CancellationToken.IsCancellationRequested)
+                        {
+                            state.TrySetCanceled(state.CancellationToken);
+                            return;
+                        }
+                        int result = Interop.libcurl.curl_easy_perform(requestHandle);
+
+                        // No more callbacks are possible once curl_easy_perform returns
+                        responseMessage.SignalHeadersReceived();
+                        responseMessage.ContentStream.SignalComplete(false);
+
+                        if (CURLcode.CURLE_OK != result)
+                        {
+#if DEBUG
+                            if (Logging.On) Logging.PrintInfo(Logging.Http, this, "Async curl_easy_perform " + Logging.GetObjectLogHash(responseMessage.ContentStream) + " failed: " + GetCurlErrorString(result));
+#endif
+                            throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                        }
+                    });
+
+                // Wait till either we got canceled or all headers have been read
+                responseMessage.WaitForHeaders(state.CancellationToken);
+
+                if (state.CancellationToken.IsCancellationRequested)
+                {
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
+                }
+
+                // Once headers are read, the response can be returned
                 state.TrySetResult(responseMessage);
+
+                await performTask;
```

Comment:
about TCS: if `await performTask;` throws then catch block will call `TrySetError`/`TrySetCancelled`. 
I don't know is this intentionally or not but looks strange.